src/object.hpp

```diff
@@ -117,88 +116,93 @@ static_assert(offsetof(object_map, ptr) == offsetof(object_array, ptr));
-
-    static auto *alloc = memory::get_default_resource();
+    // Container or non-const, non-small, string
+    return is_container(type) || type == object_type::string;
+}
 
-    auto *copy = static_cast<char *>(alloc->allocate(length, alignof(char)));
-    memcpy(copy, str, length);
+template <typename T, typename SizeType>
+inline T *alloc_helper(SizeType size, memory::memory_resource &alloc)
+    requires(std::is_unsigned_v<SizeType> && sizeof(SizeType) <= sizeof(std::size_t))
+{
+    static_assert(maxof_v<SizeType> <= maxof_v<std::size_t> / sizeof(T));
 
-    return copy;
+    return static_cast<T *>(alloc.allocate(sizeof(T) * size, alignof(T)));
 }
 
 template <typename T, typename SizeType>
-T *alloc_helper(SizeType size)
+inline void dealloc_helper(T *ptr, SizeType size, memory::memory_resource &alloc)
     requires(std::is_unsigned_v<SizeType> && sizeof(SizeType) <= sizeof(std::size_t))
-
 {
-    static auto *alloc = memory::get_default_resource();
-    // TODO add check for sizeof(T) * size
-    return static_cast<T *>(alloc->allocate(sizeof(T) * size, alignof(T)));
+    alloc.deallocate(ptr, sizeof(T) * size, alignof(T));
 }
 
 template <typename T, typename SizeType>
-inline std::pair<T *, SizeType> realloc_helper(T *data, SizeType size)
+inline std::pair<T *, SizeType> realloc_helper(
+    T *data, SizeType size, memory::memory_resource &alloc)
     requires(std::is_unsigned_v<SizeType> && sizeof(SizeType) <= sizeof(std::size_t))
 {
-    // Since allocators have no realloc interface, we're just using calloc
-    // as it'll be equivalent once allocators are supported
+    static_assert(maxof_v<SizeType> <= maxof_v<std::size_t> / sizeof(T));
+
+    constexpr std::size_t max_elements = maxof_v<SizeType>;
+
     SizeType new_size;
-    if (size > maxof_v<SizeType> / 2) [[unlikely]] {
-        new_size = maxof_v<SizeType>;
+    if (size > max_elements / 2) [[unlikely]] {
+        new_size = max_elements;
     } else {
         new_size = size * 2;
     }
 
-    static auto *alloc = memory::get_default_resource();
-
-    auto *new_data = static_cast<T *>(alloc->allocate(sizeof(T) * new_size, alignof(T)));
+    auto *new_data = static_cast<T *>(alloc.allocate(sizeof(T) * new_size, alignof(T)));
     memcpy(new_data, data, sizeof(T) * size);
 
-    alloc->deallocate(data, sizeof(T) * size, alignof(T));
+    dealloc_helper(data, size, alloc);
 
     return {new_data, new_size};
 }
 
+template <typename SizeType>
+inline char *copy_string(const char *str, SizeType length, memory::memory_resource &alloc)
+{
+    auto *copy = alloc_helper<char, uint32_t>(length, alloc);
+    memcpy(copy, str, length);
+    return copy;
+}
+
 // NOLINTNEXTLINE(misc-no-recursion)
-inline void object_destroy(object &obj)
+inline void object_destroy(object &obj, memory::memory_resource *alloc)
 {
-    static memory::memory_resource *alloc = memory::get_default_resource();
+    if (alloc->is_equal(*memory::get_default_null_resource())) {
+        return;
+    }
```

Comment:
This special case is unnecessary, no? You won't be using the null resource in normal conditions, so you shouldn't pay the penalty. In the unlikely event that you are, the code should still work correctly.

Reply:
Addressed in #420, in https://github.com/DataDog/libddwaf/pull/420/commits/bdbf8bbc892cd9e692ef61faf6f6382312fb78e1